pkg/obsservice/obslib/ingest/ingest.go

```diff
@@ -0,0 +1,182 @@
+	_ = stop.RunAsyncTask(ctx, "event ingester", func(ctx context.Context) {
+		ctx, cancel := stop.WithCancelOnQuiesce(ctx)
+		defer cancel()
+		ingestEvents(ctx, addr, db)
+	})
+}
+
+// ingestEvents subscribes to events published by the CRDB node at addr and
+// persists them to the database.
+//
+// The call blocks until the RPC is terminated by the server or ctx is canceled.
+func ingestEvents(ctx context.Context, addr string, db *pgxpool.Pool) {
+	// TODO(andrei): recover from connection errors.
+	// TODO(andrei): use certs for secure clusters.
+	conn, err := grpc.DialContext(ctx, addr,
+		grpc.WithBlock(), // block until the connection is established
+		grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		panic(fmt.Sprintf("failed to dial %s: %s", addr, err))
+	}
+	defer func() {
+		_ = conn.Close() // nolint:grpcconnclose
+	}()
+	log.Infof(ctx, "Ingesting events from %s.", addr)
+
+	c := obspb.NewObsClient(conn)
+	stream, err := c.SubscribeToEvents(ctx,
+		&obspb.SubscribeToEventsRequest{
+			Identity: "Obs Service",
+		})
+	if err != nil {
+		panic(fmt.Sprintf("SubscribeToEvents call to %s failed: %s", addr, err))
+	}
+
+	for {
+		events, err := stream.Recv()
+		if err != nil {
+			if (err != io.EOF && ctx.Err() == nil) || log.V(2) {
+				log.Infof(ctx, "event stream error: %s", err)
+			}
+			// TODO(andrei): recover from the error by trying to reestablish the
+			// connection.
+			return
+		}
+		if log.V(3) {
+			log.Infof(ctx, "received events: %s", events.String())
+		}
+		err = persistEvents(ctx, events.ResourceLogs, db)
+		if err != nil {
+			log.Errorf(ctx, "error persisting events: %s", err)
+		}
+	}
+}
+
+// persistEvents writes events to the database.
+func persistEvents(ctx context.Context, events []*otlogs.ResourceLogs, db *pgxpool.Pool) error {
+	for _, group := range events {
+		var clusterID uuid.UUID
+		var nodeID roachpb.NodeID
+		for _, att := range group.Resource.Attributes {
+			switch att.Key {
+			case obspb.ClusterID:
+				var err error
+				clusterID, err = uuid.FromString(att.Value.GetStringValue())
+				if err != nil {
+					log.Warningf(ctx, "invalid cluster ID: %s", att.Value)
+					continue
+				}
+			case obspb.NodeID:
+				nodeID = roachpb.NodeID(att.Value.GetIntValue())
+				if nodeID == 0 {
+					log.Warningf(ctx, "invalid node ID: %s", att.Value)
+					continue
+				}
+			}
+		}
+		if clusterID.Equal(uuid.UUID{}) || nodeID == 0 {
+			log.Warning(ctx, "clusterID or nodeID not set")
+			continue
+		}
+		for _, scope := range group.ScopeLogs {
+			switch scope.Scope.Name {
+			case string(obspb.EventlogEvent):
+				if err := persistEventlogEvents(ctx, scope, clusterID, nodeID, db); err != nil {
+					log.Warningf(ctx, "error persisting events: %s", err)
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+const maxEventsPerStatement = 100
+
+// persistEventlogEvents writes "eventlog" events to the database.
+func persistEventlogEvents(
```

Comment:
Any interest in making yourself an interface to write the data to persistence for later dependency injection. Feels a little crazy to be passing `pgxpool.Pool` here, but maybe that's just me. It seems like you ought to make yourself a data model and then have some interfaces to the database to interact with that model.